tee/src/lib.rs

```diff
@@ -0,0 +1,669 @@
+		///
+		/// Returns:
+		///
+		/// DispatchResultWithPostInfo
+		#[pallet::weight(T::WeightInfo::update_enclave())]
+		pub fn update_enclave(
+			origin: OriginFor<T>,
+			api_uri: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed(origin)?;
+			let enclave_id =
+				AccountEnclaveId::<T>::get(&account).ok_or(Error::<T>::NotEnclaveOwner)?;
+
+			ensure!(api_uri.len() < T::MaxUriLen::get().into(), Error::<T>::UriTooLong);
+			ensure!(api_uri.len() > T::MinUriLen::get().into(), Error::<T>::UriTooShort);
+
+			EnclaveData::<T>::mutate(enclave_id, |enclave| -> DispatchResult {
+				let enclave = enclave.as_mut().ok_or(Error::<T>::UnknownEnclaveId)?;
+				enclave.api_uri = api_uri.clone();
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::EnclaveUpdated { enclave_id, api_uri });
+			Ok(().into())
+		}
+
+		// Creates a Cluster
+		// A given cluster has list of enclaves
+		#[pallet::weight(T::WeightInfo::register_cluster())]
+		pub fn register_cluster(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			let id = ClusterIdGenerator::<T>::get();
+			let new_id = id.checked_add(1).ok_or(Error::<T>::ClusterIdOverflow)?;
+			let cluster = Cluster::new(Default::default());
+
+			ClusterData::<T>::insert(id, cluster);
+			ClusterIdGenerator::<T>::put(new_id);
+
+			Self::deposit_event(Event::ClusterAdded { cluster_id: id });
+			Ok(().into())
+		}
+
+		/// Removes a cluster
+		/// Mandate call
+		/// Cluster must be empty
+		#[pallet::weight(T::WeightInfo::unregister_cluster())]
+		pub fn unregister_cluster(
+			origin: OriginFor<T>,
+			cluster_id: ClusterId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			ClusterData::<T>::mutate(cluster_id, |cluster_opt| {
+				if let Some(cluster) = cluster_opt {
+					for enclave_id in &cluster.enclaves {
+						if EnclaveClusterId::<T>::take(enclave_id).is_none() {
+							return Err(Error::<T>::InternalLogicalError)
+						}
+					}
+					Ok(())
+				} else {
+					Err(Error::<T>::UnknownClusterId)
+				}
+			})?;
+			ClusterData::<T>::take(cluster_id);
+
+			Self::deposit_event(Event::ClusterRemoved { cluster_id });
+			Ok(().into())
+		}
+	}
+}
+
+// Helper Methods for Storage
+impl<T: Config> Pallet<T> {
+	// TODO: Replace these functions with a generic function or a Macro
+
+	/// `new_enclave_id` returns a tuple of the current enclave id and the next enclave id
+	///
+	/// Returns:
+	///
+	/// A tuple of the current enclave id and the next enclave id.
+	pub fn new_enclave_id() -> Result<(EnclaveId, EnclaveId), Error<T>> {
+		let id = EnclaveIdGenerator::<T>::get();
+		let new_id = id.checked_add(1).ok_or(Error::<T>::EnclaveIdOverflow)?;
+
+		Ok((id, new_id))
+	}
+
+	/// > This function returns a tuple of the current cluster id and the next cluster id
+	///
+	/// Returns:
+	///
+	/// A tuple of the current cluster id and the next cluster id.
+	pub fn new_cluster_id() -> Result<(ClusterId, ClusterId), Error<T>> {
```

Comment:
you can change this function to return only the new id and actually use it when getting a new id, you won't have to manually increment it and error is handled.